src/oop/object_interactions/medium/q2/MainTest.java

```diff
@@ -0,0 +1,224 @@
+            assertTrue(testClass.hasMethod(getSecretMethodName, null, String.class, new String[]{"public"}),
+                    String.format("You have modified the provided %s method method. Please revert it to the original state.", getSecretMethodName));
+            getMethodOutput = testClass.callMethod(getSecretMethodName, new String[]{"public"}, classInstance);
+            _assertEquals(secret, getMethodOutput,
+                    "You have modified the provided " + getSecretMethodName + " method on the " + className + " class. Please revert it to the original state.");
+        }
+
+    }
+
+    private static Stream<Arguments> hearSecretInputProvider() {
+        return Stream.of(
+                Arguments.of("Sam", "Sam always cheat on exams", new String[]{"Bob", "Ali"}, "Bob", "Bob has a candy"),
+                Arguments.of("John", "John is in love with Alice", new String[]{"Alice", "Bob"}, "Alice", "Alice is in love with John"),
+                Arguments.of("Kim", "Kim has broken her leg", new String[]{"Ali", "John"}, "Ali", "Ali broke Kim's leg"),
+                Arguments.of("Sam", "Sam always cheat on exams", new String[]{"Bob", "Ali"}, "John", "Bob has a candy"),
+                Arguments.of("John", "John is in love with Alice", new String[]{"Alice", "Bob", "Sam", "Neil", "Jack", "Tom", "Sara"}, "John", "Alice is in love with John"),
+                Arguments.of("Kim", "Kim has broken her leg", new String[]{"Ali", "John", "Alice", "Bob", "Sam", "Neil", "Jack", "Tom", "Sara"}, "Ali", "Ali broke Kim's leg")
+
+        );
+    }
+
+    @Test
+    public void hearSecretIsDefinedCorrectly() {
+        String incorrectMethodDefinition = "The %s method in the %s class is not defined correctly. Make sure it is declared, spelt correctly, and has the correct parameters.";
+        String incorrectModifierMessage = "The %s method in the %s class has the wrong visibility modifier.";
+        String incorrectReturnType = "The %s method in the %s class has the incorrect return type.";
+        assertTrue(testClass.hasMethod(hearSecretMethodName, new Class<?>[]{String.class, String.class}), String.format(incorrectMethodDefinition, hearSecretMethodName, className));
+        assertTrue(testClass.hasModifier(hearSecretMethodName, new Class<?>[]{String.class, String.class}, "public"), String.format(incorrectModifierMessage, hearSecretMethodName, className));
+        assertTrue(testClass.hasReturnType(hearSecretMethodName, new Class<?>[]{String.class, String.class}, Void.TYPE), String.format(incorrectReturnType, hearSecretMethodName, className));
+    }
+
+    @ParameterizedTest
+    @MethodSource("hearSecretInputProvider")
+    @Tag("dependent1")
+    public void correctHearSecretMethod(String name, String secret, String[] allFriends, String friend, String friendSecret) throws Throwable {
```

Comment:
Can you please refactor this test to see if you can add multiple secrets? We want to make sure that if they have already heard one secret, it still works on another call. You should also add a test so that it replaces a persons secret, ie they already know Bob's secret, and they hear another. I think it should be replaced (please add this to the question text if you choose to do this).